### PR TITLE
Close the data manager before replacing it

### LINKED ISSUE #598: Close the existing data manager before replacing it
When `SmartphoneStudyController` receives deployment information, it creates and configures a new `DataManager`. If a manager already exists, its reference is replaced without closing it.

The old manager remains subscribed to the study's measurement stream. For `CarpDataManager`, its SQLite buffer subscription, connectivity subscription, and upload timer also remain active. Both the old and new managers can then process the same measurements, causing duplicate SQLite rows and uploads.

This is an ownership leak: the controller owns the manager and must close it before replacing it.

Expected behavior:

- Await `DataManager.close()` before assigning a replacement.
- `CarpDataManager.close()` cancels its timer and subscriptions and closes its owned buffer.
- Each `CarpDataManager` owns a separate `DataStreamBuffer` so closing one manager cannot affect another.

### PR DESCRIPTION
**Pull request by Zeroupper**
Closes #598.

On a deployment update, `SmartphoneStudyController` created a new data manager and overwrote the reference to the old one. The old manager stayed subscribed to the study's measurement stream, so both managers handled every measurement — duplicate SQLite rows and duplicate uploads.

## `carp_mobile_sensing`

**What:** `SmartphoneStudyController` awaits `_dataManager.close()` before assigning the replacement. `AbstractDataManager.close()` awaits the subscription cancel.

**Why:** The controller owns the manager, so it must release it before dropping the reference. `close()` is awaited by that caller to know the old manager is done, so it must not complete while still subscribed.

## `carp_backend`

**What:** Added `DataStreamBuffer.detach()` and call it from `CarpDataManager.close()`. Also store and cancel the connectivity subscription, and await `buffer.initialize()`.

**Why:** `CarpDataManager.onMeasurement` is a no-op — measurements reach SQLite through the `DataStreamBuffer`'s own subscription. Closing the manager therefore did not stop it writing, and a replaced manager kept buffering measurements from the previous deployment.

`detach()` cancels that subscription but deliberately leaves the database open: the buffer is a singleton backed by a single app-wide `carp-data.db`, so closing it would break the replacement manager that is already writing to it.

The connectivity subscription was never stored in a field, so it could never be cancelled.

## Tests

Both assert the reported symptom — one measurement, one row — and each was verified to fail without its fix:

| Test | Reverting the fix gives |
|---|---|
| `carp_mobile_sensing/test/data_manager_test.dart` | `Expected: <1>, Actual: <2>` |
| `carp_backend/test/data_stream_buffer_test.dart` | `Expected: <1>, Actual: <2>` |

## Note on the third bullet in #598

> Each `CarpDataManager` owns a separate `DataStreamBuffer` so closing one manager cannot affect another.

Not implemented, deliberately. `SQLiteDataManager` hardcodes `DATABASE_NAME = 'carp-data'` with no deployment id and opens it with `singleInstance: true`, so per-instance buffers would still share one file and one handle — the isolation would be nominal only. Worse, giving each manager its own buffer and letting it close that buffer makes the old manager close the database out from under the new one, which silently drops measurements (`SQLiteDataManager.onMeasurement` returns early when the database is closed).

Real per-deployment isolation needs per-deployment database files plus a migration for the existing `carp-data.db`. That is a separate change; suggest amending that bullet on the issue.

### CODE REVIEW

File: backends/carp_backend/lib/carp_data_manager.dart [hunk 1]
```diff
@@ -307,8 +308,17 @@ class CarpDataManager extends AbstractDataManager {
   @override
   Future<void> close() async {
     uploadTimer?.cancel();
-    await uploadBufferedMeasurements();
+    uploadTimer = null;
```
Comment (jakdan99) [inline backends/carp_backend/lib/carp_data_manager.dart]: Probably should not null it, just in case

**Review by jakdan99 — CHANGES_REQUESTED** [review]

File: backends/carp_backend/lib/carp_data_manager.dart (hunk 1 above)
Comment (Zeroupper) [reply backends/carp_backend/lib/carp_data_manager.dart]: Removed both null assignments. Timer.cancel() and StreamSubscription.cancel() are sufficient and safe on repeated close() calls; configure() replaces both fields if the manager is reused. Keeping the references also avoids clearing a newly assigned subscription if configure() were to overlap with close(). The buffer lifecycle test still passes.
